pkg/cliplugins/kubestellar/get_log/cmd/cmd.go

```diff
@@ -0,0 +1,220 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/client-go/util/homedir"
+)
+
+var (
+	kubeconfig   string
+	followFlag   bool
+	linesFlag    string
+	whichArg     string
+	kubectlFlags []string
+	ksNamespace  string
+)
+
+func New(streams genericclioptions.IOStreams) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "kubestellar get_log [flags] which",
+		Short: "Get the log output from a central KubeStellar or kcp process",
+		Args:  cobra.ExactArgs(1),
+		Run:   getLog,
+	}
+
+	flags := cmd.Flags()
+	flags.StringVarP(&kubeconfig, "kubeconfig", "c", "", "Path to the kubeconfig file")
+	flags.BoolVarP(&followFlag, "follow", "f", false, "Follow the log output")
+	flags.StringVarP(&linesFlag, "lines", "n", "+0", "Number of lines to display")
+	flags.StringArrayVar(&kubectlFlags, "kubectl-flag", []string{}, "Additional kubectl flags")
+
+	if err := cmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	return cmd
+}
+
+func getLog(cmd *cobra.Command, args []string) {
+	whichArg = args[0]
+	ksNamespace = "kubestellar"
+
+	logfile := ""
+	switch whichArg {
+	case "kcp":
+		logfile = "kcp.log"
+	case "where-resolver":
+		logfile = "kubestellar-where-resolver-log.txt"
+	case "mailbox-controller":
+		logfile = "mailbox-controller-log.txt"
+	case "placement-translator":
+		logfile = "placement-translator-log.txt"
+	default:
+		fmt.Println("Argument 1 must be one of: kcp, where-resolver, mailbox-controller, placement-translator")
+		os.Exit(1)
+	}
+
+	// Build the Kubernetes clientset
+	config, err := buildConfig()
+	if err != nil {
+		fmt.Println("Error building Kubernetes config:", err)
+		os.Exit(1)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		fmt.Println("Error creating Kubernetes clientset:", err)
+		os.Exit(1)
+	}
+	namespace, err := getNamespaceNames(clientset, ksNamespace)
+	if err != nil {
+		fmt.Println("Error getting server namespace:", err)
+		os.Exit(1)
+	}
+	podName, err := getServerPodName(clientset, namespace)
+	if err != nil {
+		fmt.Println("Error getting server pod name:", err)
+		os.Exit(1)
+	}
+
+	var serverPod string
+	for {
+		if podName != "" && len(podName) > 0 {
+			serverPod = strings.Trim(string(podName), "'")
+			break
+		}
+		time.Sleep(10 * time.Second)
+	}
+
+	err = executeCommandInPod(clientset, namespace, config, podName, []string{"ls", "/home/kubestellar/ready"}, os.Stdin, os.Stdout, os.Stderr)
```

Comment:
You want this to be a test inside the loop.  The idea is to loop until the ready signal appears.

Reply:
I am not sure how to tell the difference between a failed command run and 'not ready'.  How would we tell the difference programmatically?

Reply:
I am not sure whether there _is_ a way to tell the difference.  Note that the script being replaced here does not attempt to treat those two differently.  That means that if something else is going wrong then the command will wait forever for the ready signal that never comes.  The user will have catch on an investigate.

The better approach is to recognize that Kubernetes includes a framework by which a Pod designer can specify how readiness of that Pod is tested, and use that framework.

Reply:
ok - so nothing to change here until we update the pod design.  I can create a loop in a follow-on pr when the design changes.

Reply:
The behavior currently in this PR is different from the behavior in the bash script.  The bash script will loop until the `ls` succeeds.  I think that is the better behavior to do here too.